docs/advanced/document-store/subindex.md

```diff
@@ -0,0 +1,164 @@
+
+_docs = [
+    Document(
+        text='hello world',
+        chunks=[Document(uri='image.png').load_uri_to_image_tensor()],
+    )
+    for _ in range(10)
+]
+da = DocumentArray(
+    _docs,
+    config={'n_dim': 256},
+    storage='annlite',
+    subindex_configs={'@c': {'n_dim': 128}},
+)
+```
+````
+
+The `subindex_configs` dictionary is structured in the following way:
+
+- **Keys:** Each key in `subindex_configs` is the *name* of a subindex. It has to be a valid DocumentArray access path (such as `'@.[image]'`, `'@.[image, paragraph]'`, `'@c'`, or `'@cc'`).
+
+- **Values:** Each value in `subindex_configs` is the *configuration* of a subindex. It can be any configuration that is valid for the given DocumentArray type.
+Fields that are not given in the subindex configuration will be inherited from the parent configuration.
+
+
+## Modify subindices
+
+Once a DocumentArray with subindices has been constructed, any modifications to the parent DocumentArray will automatically update the subindices.
+
+This means that you can insert, extend, delete etc. it like any other DocumentArray. For example:
+
+````{tab} Subindex with dataclass modalities
+```python
+import numpy as np
+
+# construct DocumentArry with subindices
+da = DocumentArray(
+    config={'n_dim': 256},
+    storage='annlite',
+    subindex_configs={'@.[image]': {'n_dim': 512}, '@.[paragraph]': {'n_dim': 128}},
+)
+# extend with Documents, including embeddings
+_docs = [
+    Document(MyDocument(image='image.png', paragraph='hello world')) for _ in range(10)
+]
+for d in _docs:
+    d.image.embedding = np.random.rand(512)
+    d.paragraph.embedding = np.random.rand(128)
+with da:
+    da.extend(_docs)
+```
+````
+````{tab} Subindex with chunks
+```python
+import numpy as np
+
+# construct DocumentArry with subindices
+da = DocumentArray(
+    config={'n_dim': 256},
+    storage='annlite',
+    subindex_configs={'@c': {'n_dim': 128}},
+)
+# extend with Documents, including embeddings
+_docs = [
+    Document(
+        text='hello world',
+        chunks=[Document(uri='image.png').load_uri_to_image_tensor()],
+    )
+    for _ in range(10)
+]
+for d in _docs:
+    d.embedding = np.random.rand(256)
+    d.chunks[0].embedding = np.random.rand(128)
+with da:
+    da.extend(_docs)
+```
+```` 
+
+## Search through subindices
+
+You can search through a subindex using the `on=` keyword in {meth}`~docarray.array.document.DocumentArray.find` and {meth}`~docarray.array.document.DocumentArray.match`:
+
+````{tab} Subindex with dataclass modalities
+```python
+# find best matching images using .find()
+top_image_matches = da.find(query=np.random.rand(512), on='@.[image]')
+# find best matching paragraphs using .match()
+Document(embedding=np.random.rand(128)).match(da, on='@.[paragraph]')
+```
+````
+````{tab} Subindex with chunks
+```python
+# find best matching images using .find()
+top_image_matches = da.find(query=np.random.rand(128), on='@c')
+# find best matching images using .match()
```

Comment:
```suggestion
# find best matching paragraphs using .match()
```

Reply:
The images are on the chunks level so I think it is correct the way it is. Without the use of dataclass API it is not easy to have a nesting with both images and paragraphs, so for simplicity I do everything on the images in the chunk example.

Reply:
ok, i see.